src/vecdb/handler.rs

```diff
@@ -0,0 +1,399 @@
+        let start_lines: Vec<u64> = records.iter().map(|x| x.start_line).collect();
+        let end_lines: Vec<u64> = records.iter().map(|x| x.end_line).collect();
+        let time_adds: Vec<u64> = records.iter().map(|x| x.time_added.duration_since(std::time::UNIX_EPOCH).unwrap().as_secs()).collect();
+        let model_names: Vec<String> = records.iter().map(|x| x.model_name.clone()).collect();
+        let data_batches_iter = RecordBatchIterator::new(
+            vec![RecordBatch::try_new(
+                self.schema.clone(),
+                vec![
+                    Arc::new(FixedSizeListArray::from(vectors.clone())),
+                    Arc::new(StringArray::from(window_texts.clone())),
+                    Arc::new(StringArray::from(window_text_hashes.clone())),
+                    Arc::new(StringArray::from(file_paths.clone())),
+                    Arc::new(UInt64Array::from(start_lines.clone())),
+                    Arc::new(UInt64Array::from(end_lines.clone())),
+                    Arc::new(UInt64Array::from(time_adds.clone())),
+                    Arc::new(StringArray::from(model_names.clone())),
+                ],
+            )],
+            self.schema.clone(),
+        );
+        let cache_batches_iter = RecordBatchIterator::new(
+            vec![RecordBatch::try_new(
+                self.schema.clone(),
+                vec![
+                    Arc::new(FixedSizeListArray::from(vectors)),
+                    Arc::new(StringArray::from(window_texts)),
+                    Arc::new(StringArray::from(window_text_hashes.clone())),
+                    Arc::new(StringArray::from(file_paths)),
+                    Arc::new(UInt64Array::from(start_lines)),
+                    Arc::new(UInt64Array::from(end_lines)),
+                    Arc::new(UInt64Array::from(time_adds)),
+                    Arc::new(StringArray::from(model_names)),
+                ],
+            )],
+            self.schema.clone(),
+        );
+
+        let data_res = self.data_table.add(
+            data_batches_iter, Option::from(WriteParams {
+                mode: WriteMode::Append,
+                ..Default::default()
+            }),
+        );
+
+        let cache_res = self.cache_table.add(
+            cache_batches_iter, Option::from(WriteParams {
+                mode: WriteMode::Append,
+                ..Default::default()
+            }),
+        );
+        if add_to_cache {
+            self.hashes_cache.extend(
+                records.iter().map(|x| (x.window_text_hash.clone(), x.clone())).collect::<Vec<_>>()
+            );
+            cache_res.await.unwrap();
+        }
+        self.data_table_hashes.extend(window_text_hashes);
+        data_res.await
+    }
+
+    pub async fn remove(&mut self, file_path: &PathBuf) {
+        self.cache_table.delete(
+            format!("(file_path = {})", file_path.to_str().unwrap()).as_str()
```

Comment:
Analog of SQL injection

Reply:
Made a TODO for the future, won't handle it now